### PR TITLE
Added margin loss, f1 monitoring

### PR DESCRIPTION
**Pull request by vidhishanair**

### CODE REVIEW

File: qasper_baselines/model.py [hunk 1]
```diff
@@ -154,15 +160,60 @@ def forward(
...
             )
-            loss_fn = torch.nn.CrossEntropyLoss(weight=weights)
-            evidence_loss = loss_fn(evidence_logits.view(-1, 2), evidence.view(-1))
-            self._evidence_loss(float(evidence_loss.detach().cpu()))
+            if self._use_margin_loss_for_evidence:
+                evidence_probs = torch.softmax(evidence_logits, dim=1)
+
+                positive_example_indices = (evidence == 1).nonzero()[:, 1].unsqueeze(0)
+                positive_example_scores = util.batched_index_select(evidence_probs, positive_example_indices).squeeze(0)
+                min_positive_score = torch.min(positive_example_scores).unsqueeze(0)
```
Comment (pdasigi) [inline qasper_baselines/model.py:255]: What is this value expected to be when there are no positives?

File: qasper_baselines/model.py [hunk 2]
```diff
@@ -154,15 +160,60 @@ def forward(
...
+
+                if self.training:
+                    min_positive_index = positive_example_indices[:, torch.argmin(positive_example_scores)]
+                    min_positive_prob = evidence_probs[:, min_positive_index, :]
+                    max_negative_index = negative_example_indices[:, torch.argmax(negative_example_scores)]
+                    max_negative_prob = evidence_probs[:, max_negative_index, :]
+                    if max_negative_prob < min_positive_prob:
+                        self._train_evidence_logit_thresh(max_negative_prob+(min_positive_prob-max_negative_prob)/2)
+                    else:
+                        self._train_evidence_logit_thresh(min_positive_prob+(max_negative_prob-min_positive_prob)/2)
```
Comment (pdasigi) [inline qasper_baselines/model.py:271]: This is the case where some positives are scored lower than some negatives, right? I am not sure if this should be the threshold then. Why not make the threshold always be `min(min_positive_prob, max_negative_prob)`?
